### DISCUSSION TITLE
Way to determine north/heading angle offset for projected maps

### DISCUSSION THREAD

**Discussion by chrisjaquet** [pelicanmapping/osgearth#2092] [Support Q&A] [answered]
I am trying to work out a rotation angle (or matrix) for determining the direction of North when using a projected map SRS where the direction of North varies based on geographical coordinates for the projected map. 

An example of such a projection is the Albers Conic projection (similar to the Lambert Conformal Conic):
![image](https://user-images.githubusercontent.com/1385693/178151046-2302f060-0d3b-4e49-b0eb-5bb9f08fb02a.png)
([source](https://en.wikipedia.org/wiki/List_of_map_projections#/media/File:Albers_projection_SW.jpg))
 
I have the following code snippet which successfully determines the projected coordinates (world coordinates) of a point on the projected map:
```
osgEarth::GeoPoint geoPoint(osgEarth::SpatialReference::create("wgs84"), longitude, latitude, altitude, osgEarth::ALTMODE_ABSOLUTE);
geoPoint.transformInPlace(mapSrs);

osg::Vec3d worldCoords;
geoPointPoint.toWorld(worldCoords);
```
but when I try to get the transformation matrix for this point:
```
osg::Matrixd localToWorldMatrix;
cameraPoint.createLocalToWorld(localToWorldMatrix);
```
the matrix only has a translation component, and no rotation component (i.e. `localToWorldMatrix.getRotate()` returns the identity matrix). 

Have I missed something somewhere or is it not possible with osgEarth? Any pointers or advice on where to look/poke around would be much appreciated.

OsgEarth 3.1.0
OSG 3.7.0 (748dca7, circa 2021-05-21)

**Comment by gwaldron** [comment, marked as answer]
createLocalToWorld doesn't guarantee that the +Y axis corresponds to north (as you have discovered). There isn't a "built in" way to do it but you could work it out with OSG... off the top of my head:

 * Pick two points forming a line that points north in an SRS like WGS84
 * Convert these points to your projected SRS, giving you a directional vector that will face north
 * Create a rotation matrix (or a Quat) using that vector. You should be able to multiply this into the localToWorld matrix.

Good luck!

**Reply by chrisjaquet** [reply]
Thanks, I suspected I would need to do something like this, but figured I would ask to be sure.
